be/src/runtime/routine_load/data_consumer_group.cpp

```diff
@@ -235,4 +235,203 @@ void KafkaDataConsumerGroup::actual_consume(const std::shared_ptr<DataConsumer>&
+
+    LOG(INFO) << "start consumer group: " << _grp_id << ". max time(ms): " << left_time
+              << ", batch size: " << left_bytes << ". " << ctx->brief();
+
+    // copy one
+    std::map<std::string, pulsar::MessageId> ack_offset = ctx->pulsar_info->ack_offset;
+
+    //improve performance
+    Status (PulsarConsumerPipe::*append_data)(const char* data, size_t size, char row_delimiter);
+    char row_delimiter = '\n';
+    if (ctx->format == TFileFormatType::FORMAT_JSON) {
+        append_data = &PulsarConsumerPipe::append_json;
+    } else {
+        append_data = &PulsarConsumerPipe::append_with_row_delimiter;
+        auto& per_node_scan_ranges = ctx->put_result.params.params.per_node_scan_ranges;
+
+        if (!per_node_scan_ranges.empty()) {
+            DCHECK_GE(per_node_scan_ranges.begin()->second.size(), 1);
+
+            auto& scan_range = per_node_scan_ranges.begin()->second[0].scan_range;
+            auto& params = scan_range.broker_scan_range.params;
+            row_delimiter = static_cast<char>(params.row_delimiter);
+        }
+    }
+
+    MonotonicStopWatch watch;
+    watch.start();
+    bool eos = false;
+    while (true) {
+        if (eos || left_time <= 0 || left_bytes <= 0) {
+            LOG(INFO) << "consumer group done: " << _grp_id
+                      << ". consume time(ms)=" << ctx->max_interval_s * 1000 - left_time
+                      << ", received rows=" << received_rows << ", received bytes=" << ctx->max_batch_size - left_bytes
+                      << ", eos: " << eos << ", left_time: " << left_time << ", left_bytes: " << left_bytes
+                      << ", blocking get time(us): " << _queue.total_get_wait_time() / 1000
+                      << ", blocking put time(us): " << _queue.total_put_wait_time() / 1000;
+
+            // shutdown queue
+            _queue.shutdown();
+            // cancel all consumers
+            for (auto& consumer : _consumers) {
+                consumer->cancel(ctx);
+            }
+
+            // waiting all threads finished
+            _thread_pool.shutdown();
+            _thread_pool.join();
+
+            if (!result_st.ok()) {
+                // some consumers encounter errors, cancel this task
+                return result_st;
+            }
+
+            if (left_bytes == ctx->max_batch_size) {
+                // nothing to be consumed, we have to cancel it, because
+                // we do not allow finishing stream load pipe without data
+                pulsar_pipe->cancel(Status::Cancelled("Cancelled"));
+                return Status::Cancelled("Cancelled");
+            } else {
+                DCHECK(left_bytes < ctx->max_batch_size);
+                pulsar_pipe->finish();
+                ctx->pulsar_info->ack_offset = std::move(ack_offset);
+                ctx->receive_bytes = ctx->max_batch_size - left_bytes;
+                get_backlog_nums(ctx);
```

Comment:
What's the purpose to update backlog num?

Reply:
Backlog num will be showed up as progress from FE side, so we need to update it at the end of this task.